File: python/demo_usde.py
#!/usr/bin/env python3
"""USDE Python demo — show high-level API usage."""

from usde import Network

# 1. Load a DBC database
net = Network(name="CFCAN")
ok = net.load_dbc("test_data/main.dbc")
print(f"Loaded main.dbc: {ok}, frames: {net.frame_count}")

# 2. Inspect a frame
frame_id = 0x345  # AMP_CFCAN_FrP01
info = net.frame_info(frame_id)
print(f"\n{info}")
for sig in info.signals:
    print(f"  {sig}")

# 3. Decode raw CAN bytes — one line
raw = [0x00, 0x80, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00]
decoded = net.decode_frame(frame_id, raw)
print(f"\nDecoded 0x{frame_id:X}:")
for sig in decoded:
    print(f"  {sig}")

# 4. Encode — one line
encoded = net.encode_frame(frame_id, {"AMPWorkSta": 1.0})
hex_str = " ".join(f"{b:02X}" for b in encoded)
print(f"\nEncoded AMPWorkSta=1: {hex_str}")

File: java/com/usde/DemoUsde.java
package com.usde;

import java.util.HashMap;
import java.util.Map;

/**
 * USDE Java demo — high-level API usage.
 *
 * <p>Run:
 * <pre>
 *   java -cp java/jna-5.14.0.jar;java/build
 *        -Djava.library.path=build/Release
 *        com.usde.DemoUsde
 * </pre>
 */
public class DemoUsde {

    public static void main(String[] args) {
        String dbcPath = args.length > 0 ? args[0] : "test_data/main.dbc";

        // Use try-with-resources for automatic cleanup
        try (UsdeClient client = new UsdeClient()) {

            // 1. Load DBC
            boolean ok = client.loadDbc(dbcPath);
            System.out.println("LoadDBC: " + ok + ", frames: " + client.getFrameCount());

            // 2. Prepare raw CAN bytes (frame 0x345: AMPWorkSta)
            byte[] raw = new byte[]{0x00, (byte) 0x80, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00};

            // 3. Decode — one call
            UsdeClient.DecodedSignal[] sigs = client.decodeFrame(0x345, raw);
            System.out.println("Decoded " + sigs.length + " signal(s):");
            for (UsdeClient.DecodedSignal s : sigs) {
                System.out.println("  " + s);
            }

            // 4. Encode — one call
            Map<String, Double> values = new HashMap<>();
            values.put("AMPWorkSta", 1.0);
            byte[] encoded = client.encodeFrame(0x345, values);
            System.out.print("Encoded AMPWorkSta=1: ");
            for (byte b : encoded) System.out.printf("%02X ", b);
            System.out.println();

            // 5. Round-trip verify
            UsdeClient.DecodedSignal[] rt = client.decodeFrame(0x345, encoded);
            System.out.println("Round-trip: " + rt[0]);

        } catch (Exception e) {
            System.err.println("Error: " + e.getMessage());
            e.printStackTrace();
        }
    }
}
